ambry-protocol/src/test/java/com.github.ambry.protocol/RequestResponseTest.java

```diff
@@ -133,17 +136,29 @@ public long writeTo(WritableByteChannel channel)
 }
 
 public class RequestResponseTest {
+  private Random random = new Random();
+
   private void testPutRequest(MockClusterMap clusterMap, int correlationId, String clientId, BlobId blobId,
       BlobProperties blobProperties, byte[] userMetadata, BlobType blobType, byte[] blob, int blobSize)
       throws IOException {
     int sizeInBlobProperties = (int) blobProperties.getBlobSize();
     PutRequest request = new PutRequest(correlationId, clientId, blobId, blobProperties, ByteBuffer.wrap(userMetadata),
         new ByteBufferInputStream(ByteBuffer.wrap(blob)), blobSize, blobType);
     ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-    WritableByteChannel writableByteChannel = Channels.newChannel(outputStream);
+    // Create channels in such a way that writeTo() may not be able to write out all the data at once.
+    int allocationSize = random.nextInt((int) request.sizeInBytes()) + 1;
+    int expectedWriteToCount = ((int) request.sizeInBytes() + allocationSize - 1) / allocationSize;
+    int actualWriteToCount = 0;
     while (!request.isSendComplete()) {
-      request.writeTo(writableByteChannel);
+      ByteBufferChannel channel = new ByteBufferChannel(ByteBuffer.allocate(allocationSize));
+      request.writeTo(channel);
+      ByteBuffer underlyingBuf = channel.getBuffer();
+      underlyingBuf.flip();
+      outputStream.write(underlyingBuf.array(), underlyingBuf.arrayOffset(), underlyingBuf.remaining());
```

Comment:
if that is the intent, then wouldn't doing `underlyingBuf.flip()` be wrong too since it sets the `position` to 0?